pkg/torproxy/torproxy.go

```diff
@@ -61,30 +66,80 @@ func NewTorProxyFromHostAndPort(torHost string, torPort int) (*TorProxy, error)
 func NewTorProxy() (*TorProxy, error) {
 	torClient, err := NewTorEmbedded()
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't start tor client: %w", err)
+		return nil, fmt.Errorf("couldn't start tor client: %w", err)
 	}
 
 	return &TorProxy{
 		Client: torClient,
 	}, nil
 }
 
+func (tp *TorProxy) WithRegistry(registry registry.Registry) error {
+	tp.Registry = registry
+	registryJSON, err := tp.Registry.GetJSON()
+	if err != nil {
+		return err
+	}
+
+	return tp.setRedirectsFromRegistry(registryJSON)
+}
+
 // WithRedirects modify the TorProxy struct with givend from -> to map
-func (tp *TorProxy) WithRedirects(redirects []string) error {
-	var err error
+// add the redirect URL if and only if the tor proxy doesn't know the new origin
+func (tp *TorProxy) setRedirectsFromRegistry(registryJSON []byte) error {
+	redirects, err := parseRegistryJSONtoRedirects(registryJSON)
+	if err != nil {
+		return err
+	}
+
 	for _, to := range redirects {
 		// we parse the destination upstram which should be on *.onion address
 		origin, err := url.Parse(to)
 		if err != nil {
-			err = fmt.Errorf("Failed to parse address : %v", err)
-			break
+			return fmt.Errorf("failed to parse address : %v", err)
+		}
+
+		if !tp.includesRedirect(origin) {
+			tp.Redirects = append(tp.Redirects, origin)
 		}
-		tp.Redirects = append(tp.Redirects, origin)
 	}
 
 	return err
 }
 
+func (tp TorProxy) includesRedirect(redirect *url.URL) bool {
+	for _, proxyRedirect := range tp.Redirects {
+		if proxyRedirect.Host == redirect.Host {
+			return true
+		}
+	}
+
+	return false
+}
+
+// StartAutoUpdateRedirects starts a go-routine selecting results of registry.Observe
+// return a `stop` function using to close channels and stop auto-update.
+func (tp *TorProxy) StartAutoUpdateRedirects(period time.Duration, errorHandler func (err error)) func() {
+	registryJSONchan, errors, stop := registry.Observe(tp.Registry, period)
+
+	go func() {
+		for {
+			select {
+			case json := <-registryJSONchan:
+				err := tp.setRedirectsFromRegistry(json)
+				if err != nil {
+					errorHandler(err)
+				}
+			case err := <-errors:
+				errorHandler(err)
+			}
+		}
```

Comment:
with the changes to the channel type described above, this should become a simple `for t := range greatChanName { ... }`